### ISSUE TITLE
Protractor hangs with (n) instances of webdriver still running

### ISSUE THREAD

**Issue by dborin** [angular/protractor#1130]
Shutting down selenium standalone server.
[launcher] 1 instance(s) of WebDriver still running
Shutting down selenium standalone server.
[launcher] 1 instance(s) of WebDriver still running

This sat on our Jenkins slave for 11 hours (overnight).  You can clearly see where it indicates it there is one instance, it is "Shutting down..." but then there is still one instance.

Thoughts on how I can fix this so I can utilize the excellent sharding feature?

**Comment by amitojduggal** [comment]
i believe this issue needs to be reopened, i am seeing this continuously for our project 
While running in shrded mode, with maxinstances as 10.
All goes fine until we reach the last instance.

Then it is stuck on.
[launcher] 1 instance(s) of WebDriver still running

for hours.
Is there a way to workaround this?
i am using protractor 3.0 and grunt protractor runner 3.0 as well.

**Comment by spikeybill** [comment]
Can you please post your protractor and your grunt configuration please.

**Comment by amitojduggal** [comment]
Protractor config:
exports.config = {

    baseUrl: 'https://localhost/app/',
    framework: "jasmine2",
    directConnect: true,
    // run.
    suites: {
        Suite1: 'src/e2e/tests1.js',
        Suite2: 'src/e2e/tests2.js',
        Suite3: 'src/e2e/tests3.js',
        Suite4: 'src/e2e/tests4.js',
        Suite5: 'src/e2e/test5.js',
        Suite6: 'src/e2e/tests6.js',
        Suite7: 'src/e2e/tests7.js',
      
    },
    allScriptsTimeout: 1200000,
    getPageTimeout: 500000,

    capabilities: {
        'browserName': 'chrome',
        'chromeOptions': {
            'args': ['--disable-extensions', '--incognito', ]
        }
    },
  
    //
    // See the full list at https://github.com/juliemr/minijasminenode/tree/jasmine1
    jasmineNodeOpts: {
        onComplete: null,
        isVerbose: false,
        showColors: true,
        includeStackTrace: true,
        defaultTimeoutInterval: 1800000,
    },

  };

Grunt config:
protractor: {
            options: {
                keepAlive: false,
                configFile: 'protractor.conf.js'
            },

**Comment by amitojduggal** [comment]
Protractor config:
exports.config = {

    baseUrl: 'https://localhost/app/',
    framework: "jasmine2",
    directConnect: true,
    // run.
    suites: {
        Suite1: 'src/e2e/tests1.js',
        Suite2: 'src/e2e/tests2.js',
        Suite3: 'src/e2e/tests3.js',
        Suite4: 'src/e2e/tests4.js',
        Suite5: 'src/e2e/test5.js',
        Suite6: 'src/e2e/tests6.js',
        Suite7: 'src/e2e/tests7.js',
      
    },
    allScriptsTimeout: 1200000,
    getPageTimeout: 500000,

    capabilities: {
        'browserName': 'chrome',
        'chromeOptions': {
            'args': ['--disable-extensions', '--incognito', ]
        }
    },
  
    //
    // See the full list at https://github.com/juliemr/minijasminenode/tree/jasmine1
    jasmineNodeOpts: {
        onComplete: null,
        isVerbose: false,
        showColors: true,
        includeStackTrace: true,
        defaultTimeoutInterval: 1800000,
    },

  };

Grunt config:
protractor: {
            options: {
                keepAlive: false,
                configFile: 'protractor.conf.js'
            },
chrome: {
                options: {
                    keepAlive: false,
                    args: {
                        directConnect: true,
                        capabilities: {
                            'browserName': 'chrome',
                            'shardTestFiles': true,
                            'maxInstances': 4,
                            'chromeOptions': {
                                'args': ['--disable-extensions', '--incognito', ]
                            }
                        }
                    }
                }
            }
}

Basically i use sharding parameter from grunt to override the settings in protractor.conf.js.

The same works fine with firefox all the time, but with chrome , it gets stuck on Launching instance right at the end of the execution, and the CI build has to be halted.

**Comment by spikeybill** [comment]
Can you try setting all the keepalive settings to true

**Comment by amitojduggal** [comment]
Thanks for your replies Spikey. 
i tried by setting keepAlive: true, still hitting the same issue.

seems like a bug in the way protractor handles the webdriver during shrding.
Any other workarounds i can try?

**Comment by eyphka** [comment]
same here

**Comment by spikeybill** [comment]
Just trying the obvious simple things first. I made sure all my timeout values were the same and made sure that the process running grunt had a timeout greater than the protractor / jasmine timeouts. Try it

**Comment by amitojduggal** [comment]
Thanks spikey, thats for assisting on this.
i have few questions
 from other timeouts you mean?
allScriptsTimeout: ,
getPageTimeout: 
and
defaultTimeoutInterval: 

i would like to understand what you mean by?
> the process running grunt

**Comment by spikeybill** [comment]
yes i mean allscriptsTimeout and defaultTimeoutInterval. We run grunt on windows by starting a process (.net) and this allows for a process timeout which was timing out before the tests completed,    so however you run grunt, just make sure that you don't have any timeout is all i am saying.

**Comment by amitojduggal** [comment]
Alright, so it seems like timeout can be one of the factors to it. 
the only observation i can make about timeouts is that on the CI Jenkins environment, the suites run quicker as compared to on my machine.
does that mean i should be reducing allscriptsTimeout  and defaultTimeoutInterval.
its like if suite 1 finishes in 500 second on my machine, it finishes in 300 in CI.

trying to play around with both times on my CI box, but cudnt work through yet...

**Comment by amitojduggal** [comment]
no success still with the settings, i somehow think its the same issue as reported with Chrome having issues with xvfb settings on Linux. i am not sure how to use the tullman's scripts with grunt. (from another thread"
https://github.com/angular/protractor/issues/2419

**Comment by MrFrankel** [comment]
Seems like running chrome headless doesn't hang, Id recommend running headless anyway.
